holobot/sdk/database/repositories/repository_base.py

```diff
@@ -0,0 +1,327 @@
+from abc import ABC, abstractmethod
+from collections.abc import Sequence
+from datetime import datetime
+from enum import Enum
+from typing import Any, Callable, Generic, TypeVar, cast
+
+from asyncpg.connection import Connection
+
+from holobot.sdk import Lazy
+from holobot.sdk.database import DatabaseManagerInterface
+from holobot.sdk.database.exceptions import DatabaseError
+from holobot.sdk.database.queries import (
+    CompiledQuery, ICompileableQueryPartBuilder, ISupportsPagination, Query, WhereBuilder
+)
+from holobot.sdk.database.queries.enums import Equality
+from holobot.sdk.database.statuses import CommandComplete
+from holobot.sdk.database.statuses.command_tags import DeleteCommandTag, UpdateCommandTag
+from holobot.sdk.queries import PaginationResult
+from holobot.sdk.utils import UTC, set_time_zone
+from holobot.sdk.utils.dataclass_utils import ArgumentInfo, get_argument_infos
+from .entity import Entity
+from .irepository import IRepository
+
+TIdentifier = TypeVar("TIdentifier")
+TRecord = TypeVar("TRecord", bound=Entity)
+TModel = TypeVar("TModel")
+
+_ID_FIELD_NAME = "id"
+
+class RepositoryBase(
+    ABC,
+    Generic[TIdentifier, TRecord, TModel],
+    IRepository[TIdentifier, TModel]
+):
+    """Abstract base class for a repository."""
+
+    @property
+    def column_names(self) -> tuple[str, ...]:
+        """Gets the names of the table's columns.
+
+        This is a cached value.
+
+        :return: The names of the table's columns.
+        :rtype: tuple[str, ...]
+        """
+
+        return self.__column_names.value
+
+    @property
+    @abstractmethod
+    def record_type(self) -> type[TRecord]:
+        """Gets the type of the record.
+
+        This is required, because Python currently doesn't support
+        the runtime resolution of generic type arguments.
+
+        :return: The type of the record.
+        :rtype: type[TRecord]
+        """
+
+    @property
+    @abstractmethod
+    def table_name(self) -> str:
+        ...
+
+    def __init__(
+        self,
+        database_manager: DatabaseManagerInterface
```

Comment:
Use this opportunity to rename `DatabaseManagerInterface` to `IDatabaseManager`.